openrtb-json/src/main/scala/com/powerspace/openrtb/json/BidSerde.scala

```diff
@@ -1,10 +1,42 @@
 package com.powerspace.openrtb.json
 
+import com.google.openrtb.BidResponse.SeatBid
 import com.google.openrtb.BidResponse.SeatBid.Bid.AdmOneof
 import com.google.openrtb._
-import io.circe.Decoder
+import io.circe.{Decoder, DecodingFailure, Json}
+import io.circe.parser._
 
+/**
+  * Serialize and Deserialize an OpenRTB Bid
+  */
 object BidSerde {
+
+  val nativeObject: Decoder[Json] = cursor => cursor.downField("native").as[Json]
+  val nativeDecoder: Decoder[NativeResponse] = nativeObject.emapTry(NativeSerde.decoder.decodeJson(_).toTry)
+
+  implicit class LoggableEither[L, R](tryInstance: Either[L, R]) {
+    def doOnError(action: (L) => Unit): Either[L, R] = {
+      tryInstance.fold(throwable => action(throwable), _ => Unit)
+      tryInstance
+    }
+  }
+
+  /**
+    * adm field can contain either a serialized native
+    * response string or a whole native response object
+    * @todo simplify this code
+    */
+  private implicit val admOneOfDecoder: Decoder[SeatBid.Bid.AdmOneof] =
+    cursor => for {
+      adm <- cursor.as[Option[String]]
+      parsed = adm.toRight(DecodingFailure("Unparsable tag adm.", List())).flatMap(parse)
```

Comment:
Do we really need to trigger the unparsable adm tag error, if for instance the incoming request is not native.

Reply:
If the *adm* is not native but valid it triggers the error at line 33, doesn't it?